src/nifs/protogalaxy/poly.rs

```diff
@@ -0,0 +1,178 @@
+use std::num::NonZeroU32;
+
+use ff::Field;
+use halo2curves::CurveAffine;
+use itertools::Itertools;
+use rayon::prelude::*;
+use tracing::*;
+
+use crate::{
+    fft,
+    plonk::{self, eval, GetChallenges, GetWitness, PlonkStructure},
+    polynomial::{lagrange, univariate::UnivariatePoly},
+};
+
+use super::pow_i;
+
+#[derive(Debug, thiserror::Error, PartialEq, Eq, Clone)]
+pub enum Error {
+    #[error(transparent)]
+    Eval(#[from] eval::Error),
+    #[error("pow_i iterator ended before witnesses")]
+    PowiEndedBeforeWitnesses,
+}
+
+/// Analog of [`itertools::structs::MultiProduct`], but:
+/// - without [`Clone`] require
+/// - without logic for diff len of iterators
+struct MultiProduct<Item, Iter>
```

Comment:
can we avoid add another layer of abstraction?

Reply:
This is not an abstraction level, but a helper struct, since the same struct from itertools is a bit off for us (more complext & want clone for Item)

The only thing I can do is to separate it into a separate utility crate and take it out of sirius and just import it here, but it is very small for this approach